castai/resource_role_bindings.go

```diff
@@ -0,0 +1,408 @@
+										DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+											return strings.EqualFold(oldValue, newValue)
+										},
+									},
+									FieldRoleBindingsSubjectUserID: {
+										Type:        schema.TypeString,
+										Optional:    true,
+										Computed:    true,
+										Description: fmt.Sprintf("Optional, required only if `%s` is `%s`.", FieldRoleBindingsSubjectKind, RoleBindingSubjectKindUser),
+									},
+									FieldRoleBindingsSubjectServiceAccountID: {
+										Type:        schema.TypeString,
+										Optional:    true,
+										Computed:    true,
+										Description: fmt.Sprintf("Optional, required only if `%s` is `%s`.", FieldRoleBindingsSubjectKind, RoleBindingSubjectKindServiceAccount),
+									},
+									FieldRoleBindingsSubjectGroupID: {
+										Type:        schema.TypeString,
+										Optional:    true,
+										Computed:    true,
+										Description: fmt.Sprintf("Optional, required only if `%s` is `%s`.", FieldRoleBindingsSubjectKind, RoleBindingSubjectKindGroup),
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceRoleBindingsRead(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	organizationID := data.Get(FieldRoleBindingsOrganizationID).(string)
+	if organizationID == "" {
+		var err error
+		organizationID, err = getDefaultOrganizationId(ctx, meta)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf("getting default organization: %w", err))
+		}
+	}
+	roleBindingID := data.Id()
+
+	client := meta.(*ProviderConfig).api
+
+	roleBinding, err := getRoleBinding(client, ctx, organizationID, roleBindingID)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("getting role binding for read: %w", err))
+	}
+
+	if err := assignRoleBindingData(roleBinding, data); err != nil {
+		return diag.FromErr(fmt.Errorf("assigning role binding data for read: %w", err))
+	}
+
+	return nil
+}
+
+func resourceRoleBindingsCreate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	organizationID := data.Get(FieldRoleBindingsOrganizationID).(string)
+
+	client := meta.(*ProviderConfig).api
+
+	subjects, err := convertSubjectsToSDK(data)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	scope := convertScopeToSDK(data)
+
+	resp, err := client.RbacServiceAPICreateRoleBindingsWithResponse(ctx, organizationID, sdk.RbacServiceAPICreateRoleBindingsJSONRequestBody{
+		{
+			Definition: sdk.CastaiRbacV1beta1RoleBindingDefinition{
+				RoleId:   data.Get(FieldRoleBindingsRoleID).(string),
+				Scope:    scope,
+				Subjects: &subjects,
+			},
+			Description: lo.ToPtr(data.Get(FieldRoleBindingsDescription).(string)),
+			Name:        data.Get(FieldRoleBindingsName).(string),
+		},
+	})
+
+	if err := sdk.CheckOKResponse(resp, err); err != nil {
+		return diag.FromErr(fmt.Errorf("create role binding: %w", err))
+	}
+
+	if len(*resp.JSON200) == 0 {
```

Comment:
would we ever get to this point? `sdk.CheckOKResponse` isn't gonna handle the 5XX / 4XX cases?

Reply:
Theoretically not. I wanted to be double sure that we don't run into `index out of range` in line 211